parity-crypto/src/publickey/ecies.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Functions for ECIES scheme encryption and decryption
+
+use ethereum_types::H128;
+use super::{Error, Random, Generator, Public, Secret, ecdh};
+use crate::{aes, digest, hmac, is_equal};
+
+const ENC_VERSION: u8 = 0x04;
+
+/// Encrypt a message with a public key, writing an HMAC covering both
+/// the plaintext and authenticated data.
+///
+/// Authenticated data may be empty.
+pub fn encrypt(public: &Public, auth_data: &[u8], plain: &[u8]) -> Result<Vec<u8>, Error> {
+	let r = Random.generate()?;
+	let z = ecdh::agree(r.secret(), public)?;
+	let mut key = [0u8; 32];
+	kdf(&z, &[0u8; 0], &mut key);
+
+	let ekey = &key[0..16];
+	let mkey = hmac::SigKey::sha256(&digest::sha256(&key[16..32]));
+
+	let mut msg = vec![0u8; 1 + 64 + 16 + plain.len() + 32];
+	msg[0] = ENC_VERSION;
+	{
+		let result_msg = &mut msg[1..];
+		result_msg[0..64].copy_from_slice(r.public().as_bytes());
+		let iv = H128::random();
+		result_msg[64..80].copy_from_slice(iv.as_bytes());
+		{
+			let cipher = &mut result_msg[(64 + 16)..(64 + 16 + plain.len())];
+			aes::encrypt_128_ctr(ekey, iv.as_bytes(), plain, cipher)?;
+		}
+		let mut hmac = hmac::Signer::with(&mkey);
+		{
+			let cipher_iv = &result_msg[64..(64 + 16 + plain.len())];
+			hmac.update(cipher_iv);
+		}
+		hmac.update(auth_data);
+		let sig = hmac.sign();
+		result_msg[(64 + 16 + plain.len())..].copy_from_slice(&sig);
+	}
+	Ok(msg)
+}
+
+/// Decrypt a message with a secret key, checking HMAC for ciphertext
+/// and authenticated data validity.
+pub fn decrypt(secret: &Secret, auth_data: &[u8], encrypted: &[u8]) -> Result<Vec<u8>, Error> {
+	const META_LEN: usize = 1 + 64 + 16 + 32;
+	let enc_version = encrypted[0];
+	if encrypted.len() < META_LEN || enc_version < 2 || enc_version > 4 {
+		return Err(Error::InvalidMessage); //invalid message: publickey
```

Comment:
Not sure I understand the comment here. Can you clarify?

Reply:
Comment removed